pkg/jobs/adopt.go

```diff
@@ -53,6 +54,30 @@ const claimQuery = `
     LIMIT $3
 RETURNING id;`
 
+func (r *Registry) maybeDumpTrace(
+	resumerCtx context.Context, resumer Resumer, jobID, traceID int64, jobErr error,
+) {
+	if _, ok := resumer.(TraceableJob); !ok || r.td == nil {
+		return
+	}
+	dumpMode := traceableJobDumpTraceMode.Get(&r.settings.SV)
+	if dumpMode == int64(noDump) {
+		return
+	}
+
+	// If the job has failed, and the dump mode is set to anything except noDump,
+	// then we should dump the trace.
+	if jobErr != nil && resumerCtx.Err() == nil {
+		r.td.MaybeDump(resumerCtx, strconv.Itoa(int(jobID)), traceID, r.ex)
+	}
+
+	// If the job has succeeded, and the dump mode is set to dumpOnTerminal, then
```

Comment:
Based on the internal usage, `terminal` would mean success, fail, or cancel. Should canceled jobs be included here?

Reply:
@pbardea can I have you take another pass at the job-related logic (second commit). Things that have changed:

- Cluster setting modes.
- Dump onFail or vs every other state.
- Restructured the tests in jobs_test.go. I couldn't think of a cleaner way but am open to suggestions.